proxy/server/server.go

```diff
@@ -270,30 +281,91 @@ func (s *T) newTLSCfg(hostCfgs map[engine.HostKey]engine.Host) (*tls.Config, err
-			r, err := s.stapler.StapleHost(&hostCfg)
+		if hostCfg.Settings.KeyPair != nil {
+			c := hostCfg.Settings.KeyPair
+			if hostCfg.Settings.AutoCert != nil {
+				log.Warningf("Host %v has a KeyPair and AutoCert enabled. KeyPair takes precedence. Autocert generation disabled.", hostCfg.Name)
+			}
+			keyPair, err := tls.X509KeyPair(c.Cert, c.Key)
+			if err != nil {
+				return nil, err
+			}
+			if hostCfg.Settings.OCSP.Enabled {
+				log.Infof("%v OCSP is enabled for %v, resolvers: %v", s, hostCfg, hostCfg.Settings.OCSP.Responders)
+				r, err := s.stapler.StapleHost(&hostCfg)
+				if err != nil {
+					log.Warningf("%v failed to staple %v, error %v", s, hostCfg, err)
+				} else if r.Response.Status == ocsp.Good || r.Response.Status == ocsp.Revoked {
+					keyPair.OCSPStaple = r.Staple
+				} else {
+					log.Warningf("%s got undefined status from OCSP responder: %v", s, r.Response.Status)
+				}
+			}
+			pairs[hostCfg.Name] = keyPair
+		} else if hostCfg.Settings.AutoCert != nil {
+			ac := hostCfg.Settings.AutoCert
+			if s.autoCertMgrs == nil {
+				s.autoCertMgrs = make([]*autocert.Manager, 0, 0)
+			}
+
+			// Each host gets its own Autocert Manager - this allows individual
+			// certs to use different autocert authorities, as well as auth keys
+			autoCertMgr := &autocert.Manager{
+				Prompt:     autocert.AcceptTOS,
+				Cache:      s.autoCertCache,
+				HostPolicy: autocert.HostWhitelist(hostCfg.Name),
+				Email:      ac.Email,
+			}
+
+			if ac.RenewBeforeSeconds > 0 {
+				autoCertMgr.RenewBefore = time.Duration(ac.RenewBeforeSeconds) * time.Second
+			}
+
+			if ac.Key != "" || ac.DirectoryURL != "" {
+				autoCertMgr.Client = &acme.Client{
+					DirectoryURL: ac.DirectoryURL,
+				}
+
+				if ac.Key != "" {
+					block, _ := pem.Decode([]byte(ac.Key))
+					if block == nil {
+						log.Errorf("Autocert Key PEM Block for Host %s is invalid.", hostCfg.Name)
+					} else if block.Type == "RSA PRIVATE KEY" {
+						rsaPrivateKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+						if err != nil {
+							autoCertMgr.Client.Key = rsaPrivateKey
+						} else {
+							log.Errorf("Error parsing Autocert Key block of type %s, for Host %s, as an RSA Private Key: %v", block.Type, hostCfg.Name, err)
+						}
+					} else if block.Type == "EC PRIVATE KEY" {
+						ecPrivateKey, err := x509.ParseECPrivateKey(block.Bytes)
+						if err != nil {
+							autoCertMgr.Client.Key = ecPrivateKey
+						} else {
+							log.Errorf("Error parsing Autocert Key block of type %s, for Host %s, as an EC Private Key: %v", block.Type, hostCfg.Name, err)
+						}
+					} else {
+						log.Errorf("AutoCert Private Key for Host %s is of unrecognized type: %s. Supported types"+
+							"are RSA PRIVATE KEY and EC PRIVATE KEY.", hostCfg.Name, block.Type)
+					}
+				}
+			}
+
+			cert, err := autoCertMgr.GetCertificate(&tls.ClientHelloInfo{ServerName: hostCfg.Name})
```

Comment:
How is it possible to get a certificate here if the server is not yet listening with this tls.Config? Who would respond to HTTP challenge?

Reply:
I think the autocert manager within itself opens up a temporary endpoint. Somehow I've always found it able to get a cert each time I tried this.

Reply:
Actually none of my UTs really test for that. I'm removing it.